### PR TITLE
SQL Kernel Improvements/Removing Spark Code from Core/Attach to Changes

### PR DESCRIPTION
**Pull request by chlafreniere**
Sorry for the largish PR. This implements #3794. Here's what this PR covers:

- Jupyter/Spark related code has been taken out of ADS, and will (for now) live in the sql-next extension; this implies that everything in core is now more generic, as you will see below (example: use serverName property in connection object instead of checking for options['host'])
- Attach to is functional for SQL kernel, filters out non-MSSQL connections (yes, I know, we really should be filtering out 'everything that doesn't implement the DMP'). @kevcunnane any assistance with that would be most helpful
- The concept of StandardKernels is now leveraged; when a SessionManager exists and has been started, we can query that for its specs, but before a SessionManager exists, we can now map a kernel's name to both its notebook provider and to its connection provider. This comes in handy when, for example, we're opening a notebook with saved kernel information; we can then set the correct notebook provider ID before starting a session. Mementos can be leveraged here in the future, if so desired.
- Implement interrupt() so that stopping a query using the SQL kernel doesn't throw a weird exception
- Improved current grid support (i.e. stop right aligning everything)

Note, with this PR, there is now a behavior change in how the default kernel is set. Here's how the logic now works:
- If a kernel was saved previously, we use this. If an active connection wasn't found, we default to the Attach To behavior where we have "Select Connection" and "Add New Connection". We no longer will fallback to Python3.
- SQL is the default if a kernel wasn't saved, or if a default kernel is not passed in (note: this means that PySpark3 will not be the default kernel any longer for HADOOP_KNOX connections; with the unified connection work, this should make more sense)
- In cases where a kernel is passed in (currently only "Analyze in Notebook", we use that kernel). Analyze in Notebook still defaults to PySpark3.

There is also a corresponding PR on the extension side that I'll post later today. Hopefully that will make things clearer. Looking forward to the multitude of comments and sanity checking my design :).

### CODE REVIEW

File: src/sql/parts/notebook/models/notebookModel.ts [hunk 1]
```diff
@@ -207,6 +208,14 @@ export class NotebookModel extends Disposable implements INotebookModel {
 		return this._onClientSessionReady.event;
 	}
 
+	public getApplicableConnectionProviderIds(kernelDisplayName: string): string[] {
+		let ids = [];
+		if (kernelDisplayName) {
+			ids = this._kernelDisplayNameToConnectionProviderIds.get(kernelDisplayName);
+		}
+		return !ids ? [] : ids;
```
Comment (rajmusuku) [inline src/sql/parts/notebook/models/notebookModel.ts]: Can't we return 'ids' directly without a condition since it was initialized already? Do we get undefined value ?

File: src/sql/parts/notebook/models/cell.ts [hunk 2]
```diff
@@ -371,17 +356,15 @@ export class CellModel implements ICellModel {
...
-					this._language = languageInfo.name;
+			if (languageInfo.codemirror_mode) {
+				let codeMirrorMode: nb.ICodeMirrorMode = <nb.ICodeMirrorMode>(languageInfo.codemirror_mode);
+				if (codeMirrorMode && codeMirrorMode.name) {
+					this._language = codeMirrorMode.name;
 				}
-			}
-			else if (languageInfo.mimetype) {
+			} else if (languageInfo.mimetype) {
 				this._language = languageInfo.mimetype;
```
[the remark this answers is not in the document]
Comment (chlafreniere) [reply src/sql/parts/notebook/models/cell.ts]: This is by design actually. We handle the mimetype case below. Here's an example:

The PySpark3 kernel will inexplicably return "PySpark3" as its 'name' in languageInfo, but "x-python" as the mime type.

File: src/sql/parts/notebook/models/clientSession.ts [hunk 3]
```diff
@@ -60,9 +58,8 @@ export class ClientSession implements IClientSession {
 		this._kernelChangeCompleted = new Deferred<void>();
 	}
 
-	public async initialize(connection?: NotebookConnection): Promise<void> {
+	public async initialize(connection?: IConnectionProfile): Promise<void> {
 		try {
-			this._kernelConfigActions.push((kernelName: string) => { return this.runTasksBeforeSessionStart(kernelName); });
 			this._connection = connection;
```
[the remark this answers is not in the document]
Comment (chlafreniere) [reply src/sql/parts/notebook/models/clientSession.ts]: Actually don't need this anymore. Removed.

File: src/sql/parts/notebook/models/notebookModel.ts (hunk 1 above)
Comment (chlafreniere) [reply src/sql/parts/notebook/models/notebookModel.ts]: I'd rather return an empty array than return undefined (so that the caller doesn't need to check undefined), so that's why I'm doing this here.

File: src/sql/parts/notebook/models/clientSession.ts [hunk 4]
```diff
@@ -114,7 +111,7 @@ export class ClientSession implements IClientSession {
 		} catch (err) {
 			// TODO move registration
```
[the remark this answers is not in the document]
Comment (chlafreniere) [reply src/sql/parts/notebook/models/clientSession.ts]: @kevcunnane can you help clarify here? Not exactly sure what that comment refers to.

File: src/sql/parts/notebook/models/notebookContexts.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	public static async getContextsForKernel(connectionService: IConnectionManagementService, connProviderIds: string[], kernelChangedArgs?: nb.IKernelChangedArgs, profile?: IConnectionProfile): Promise<IDefaultConnection> {
+		let connections: IDefaultConnection = this.DefaultContext;
+		if (!profile) {
+			if (!kernelChangedArgs || !kernelChangedArgs.newValue ||
+				(kernelChangedArgs.oldValue && kernelChangedArgs.newValue.id === kernelChangedArgs.oldValue.id)) {
+				// nothing to do, kernels are the same or new kernel is undefined
+				return connections;
+			}
+		}
+		if (kernelChangedArgs && kernelChangedArgs.newValue && kernelChangedArgs.newValue.name) {
```
[the remark this answers is not in the document]
Comment (chlafreniere) [reply src/sql/parts/notebook/models/notebookContexts.ts]: Reworked this. Thanks.

File: src/sql/parts/notebook/models/notebookContexts.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+		if (specs) {
+			defaultKernel = specs.kernels.find((kernel) => kernel.name === specs.defaultKernel);
+			if (savedKernelInfo) {
+				foundSavedKernelInSpecs = specs.kernels.find((kernel) => kernel.name === savedKernelInfo.name);
+			}
+		}
+		if (specs && connectionInfo) {
+			// set default kernel to default kernel not saved previously
+			// otherwise, set default to kernel info loaded from existing file
+			defaultKernel = !foundSavedKernelInSpecs ? defaultKernel : foundSavedKernelInSpecs;
```
[the remark this answers is not in the document]
Comment (chlafreniere) [reply src/sql/parts/notebook/models/notebookContexts.ts]: Good catch. We've had lots of this left over from SparkMagicContexts. Let's see how we can make it faster :)

File: src/sql/parts/notebook/models/modelInterfaces.ts [hunk 7]
```diff
@@ -418,16 +429,15 @@ export interface INotebookModelOptions {
...
 	providerId: string;
+	standardKernels: IStandardKernelWithProvider[];
+	defaultKernel: nb.IKernelSpec;
 
 	notificationService: INotificationService;
 	connectionService: IConnectionManagementService;
+	capabilitiesService: ICapabilitiesService;
 }
 
 // TODO would like to move most of these constants to an extension
```
Comment (corivera) [inline src/sql/parts/notebook/models/modelInterfaces.ts]: Seems like this TODO is resolved here

File: src/sql/parts/notebook/models/cell.ts [hunk 8]
```diff
@@ -239,8 +236,8 @@ export class CellModel implements ICellModel {
 				let result = output as nb.IDisplayResult;
 				if (result && result.data && result.data['text/html']) {
 					let nbm = (this as CellModel).options.notebook as NotebookModel;
-					if (nbm.hadoopConnection) {
-						let host = nbm.hadoopConnection.host;
+					if (nbm.activeConnection) {
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/cell.ts]: Minor: we normally call it 'model' not nbm

File: src/sql/parts/notebook/models/clientSession.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevcunnane) [reply src/sql/parts/notebook/models/clientSession.ts]: So the challenge is this error (with status code 501) is somewhat specialized to Jupyter. Given the value add, I don't see us moving this out (we will just keep special casing Jupyter) but sometime waaaay in the future would be nice to move to an extension point, like when we have special case number 3 needing to be fixed.

File: src/sql/parts/notebook/models/notebookContexts.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+			defaultConnection: defaultConnection,
+			otherConnections: [defaultConnection]
+		};
+	}
+
+	private static get LocalContext(): IDefaultConnection {
+		let localConnection: ConnectionProfile = <any>{
+			providerName: NotebookContexts.MSSQL_PROVIDER,
+			id: '-1',
+			serverName: localize('localhost', 'Localhost')
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookContexts.ts]: I think `localhost` lowercased is a valid connection (e.g. server called localhost). I think "Not Connected" would be more accurate??

Comment (kevcunnane) [reply src/sql/parts/notebook/models/notebookContexts.ts]: At least in the SQL case... I know that for PySpark etc. it might be different. We should ask others about this.

File: src/sql/parts/notebook/models/notebookContexts.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+		return {
+			// default context if no other contexts are applicable
+			defaultConnection: localConnection,
+			otherConnections: [localConnection]
+		};
+	}
+
+	/**
+	 * Get all of the applicable contexts for a given kernel
+	 * @param apiWrapper ApiWrapper
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookContexts.ts]: These params are not accurate, please remove or update

File: src/sql/parts/notebook/models/notebookContexts.ts [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+	 * @param specs kernel specs (comes from session manager)
+	 * @param connectionInfo connection profile
+	 * @param savedKernelInfo kernel info loaded from
+	 */
+	public static getDefaultKernel(specs: nb.IAllKernels, connectionInfo: IConnectionProfile, savedKernelInfo: nb.IKernelInfo, notificationService: INotificationService): nb.IKernelSpec {
+		let savedKernel: nb.IKernelSpec;
+		let defaultKernel: nb.IKernelSpec;
+		if (specs) {
+			defaultKernel = specs.kernels.find((kernel) => kernel.name === specs.defaultKernel);
+			if (savedKernelInfo) {
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookContexts.ts]: Maybe comment this flow, it's hard to follow what's going on here.

File: src/sql/parts/notebook/models/notebookContexts.ts [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+	/**
+	 * Get all active contexts and sort them
+	 * @param apiWrapper ApiWrapper
+	 * @param profile current connection profile
+	 */
+	public static async getActiveContexts(connectionService: IConnectionManagementService, connProviderIds: string[], profile: IConnectionProfile): Promise<IDefaultConnection> {
+		let defaultConnection: ConnectionProfile = NotebookContexts.DefaultContext.defaultConnection;
+		let activeConnections: ConnectionProfile[] = await connectionService.getActiveConnections();
+		// If no connections exist, only show 'n/a'
+		if (activeConnections && activeConnections.length > 0) {
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookContexts.ts]: Maybe doc why we have this set of decisions. E.g. pseudocode on what's meant to happen.

File: src/sql/parts/notebook/models/notebookModel.ts [hunk 13]
```diff
@@ -380,45 +395,48 @@ export class NotebookModel extends Disposable implements INotebookModel {
...
+						if (kernel.info) {
+							this.updateLanguageInfo(kernel.info.language_info);
+						}
+					}, err => undefined);
+				}).catch((err) => {
+					this.notifyError(localize('changeKernelFailed', 'Failed to change kernel: {0}', notebookUtils.getErrorMessage(err)));
+					// TODO should revert kernels dropdown
+				});
+		}
+		return Promise.resolve();
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookModel.ts]: What's this case? Is it initiating a new session, and will the kernel ever be set?

File: src/sql/parts/notebook/models/notebookModel.ts [hunk 14]
```diff
@@ -499,6 +523,7 @@ export class NotebookModel extends Disposable implements INotebookModel {
 	public async handleClosed(): Promise<void> {
 		try {
 			if (this._activeClientSession) {
+				await this._activeClientSession.ready;
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookModel.ts]: If session ready is ever rejected this will cause problems. If it's not ready, can we do something to still shut it down / end this? 
- At a minimum, try/catch the ready call
- consider if awaiting this will potentially take a long time, and use a timeout to just skip or hard-shutdown the session in that case (e.g. if it's longer than 10 seconds)

File: src/sql/parts/notebook/models/notebookContexts.ts (hunk 9 above)
Comment (chlafreniere) [reply src/sql/parts/notebook/models/notebookContexts.ts]: Heh, so... Fun fact, even though localConnection's provider is MSSQL, this code can only possibly be hit when a connection provider id isn't provided by a kernel.

The reason why MSSQL is the provider name is that I needed to put... _something_ there, and didn't want any Jupyter references, so that's the story there.

File: src/sql/parts/notebook/models/notebookModel.ts (hunk 13 above)
Comment (chlafreniere) [reply src/sql/parts/notebook/models/notebookModel.ts]: The case is "no active client session exists". I can add an error message if you think it's helpful.
